README.md
# FiveInRow


ChangeLog.md
# Changelog for FiveInRow

## Unreleased changes


src/test.hs
module Test where

import Control.Monad (sequence_, mapM_)
import Data.Array.IO --( getElems, readArray, writeArray, MArray(newArray), IOArray )
import Data.Foldable (Foldable(foldl'))
import Data.Time.Clock ( getCurrentTime, UTCTime(utctDayTime) )
import Data.Maybe ( fromJust, isJust )
import System.Random ( initStdGen, uniformR, StdGen )
import System.IO ( hFlush, stdout )
import Control.Concurrent ( threadDelay )
import Data.List (sort)
import Consul
import Draw
import Lib2

t1 = do
  t <- newTable
  puts t [45,55,65,75] 'o'
  putStr clrscr
  drawTable t
  putStrLn $ norm ++ showCur
  return t

t2 = do
  t <- newTable
  puts t [01,03,05,07] 'x'
  putStr clrscr
  drawTable t
  putStrLn $ norm ++ showCur
  return t

t3 = do
  t <- newTable
  puts t [45,55,65,95] 'x'
  putStr clrscr
  drawTable t
  putStrLn $ norm ++ showCur
  return t

__pwp prompt x =  print $ prompt ++ show x

-- findSamples --------- поиск образцов 

test0 :: IO Table -> IO ()
test0 table = do
  t <- table
  matr <- getElems t 
  let samplesO = findSamplesT 'o' matr
  __pwp "samples finded by o " samplesO
  let samplesX = findSamplesT 'x' matr
  __pwp "samples finded by x " samplesX
  

-- priceTableAfterStep -- оценка таблицы после сделанного хода 

test1 table = do
  t <- table
  priceX <- priceTableAfterStep 'x' t
  __pwp "ater X step " priceX
  priceO <- priceTableAfterStep 'o' t
  __pwp "ater O step " priceO

-- selStepsOnTable -- выбор походящих ходов из заданной таблицы 

test2 table = do
  t <- table
  steps <- selStepsOnTable 'o' t
  __pwp "steps on table " steps

---------------------------------------------

test3 table = do
  t <- table
  steps <- nextSteps 'o' t 1 
  __pwp "steps on table " steps



src/Lib.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Use putStrLn" #-}
{-# HLINT ignore "Eta reduce" #-}
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
{-# HLINT ignore "Use mapM_" #-}
{-# HLINT ignore "Use first" #-}


module Lib (
  _size, interpolation, nextStep, put, puts, isCellEmpty, newTable, whoWon,  Table, Pos
 ) where

import Control.Monad (sequence_, mapM_)
import Data.Array.IO --( getElems, readArray, writeArray, MArray(newArray), IOArray )
import Data.Foldable (Foldable(foldl'))
import Data.Time.Clock ( getCurrentTime, UTCTime(utctDayTime) )
import Data.Maybe ( fromJust, isJust )
import System.Random ( initStdGen, uniformR, StdGen )
import Consul
import System.IO ( hFlush, stdout )
import Control.Concurrent ( threadDelay )
import Data.List (sort)


type Table = IOArray Int Val
type Pos = (Int, Int)
type Matr = [Val]
type Sample = [Val]
type Price = Int
type Val = Char    -- 'x', 'o', ' '
_size = 10

get :: Table -> Pos -> IO Val
get table (r, c) = readArray table (r * _size + c)

put :: Table -> Pos -> Val -> IO ()
put table (r, c) = writeArray table (r * _size + c)

puts :: Table -> [Int] -> Val -> IO ()
puts table ns v = mapM_ (\n -> put table (divMod n 10) v) ns

newTable :: IO Table
newTable = newArray (0, _size^2 - 1) ' '

-- UTILS ------------------------------------------------

interpolation ((r1, c1), (r2, c2))
  | r1 == r2  = [(r1, c) | c <- [c1..c2] ]
  | c1 == c2  = [(r, c1) | r <- [r1..r2] ]
  | otherwise = [(r, c)  | (r, c) <- zip [r1..r2] [c1, c1+dCol..c2]]
  where
    dCol = signum (c2 - c1)

insertSepByN n [] sep = []
insertSepByN n xs sep = let (zs, rs) = splitAt n xs
               in zs ++ [sep] ++ insertSepByN n rs sep

---------------------------------------------------------
_pricedSamples :: [([Sample], Price)]
_pricedSamples = [
  ([" oooo", "o ooo", "oo oo", "ooo o", "oooo "], 1000),
  ([" xxxx", "x xxx", "xx xx", "xxx x", "xxxx "], 999),

  ([" ooo "], 100),
  ([" xxx "], 99),

  ([" ooo", "o oo", "oo o", "ooo "], 50),
  ([" xxx", "x xx"," xx x", "xxx "], 49),

   ([" oo", "o o", "oo "], 2),
   ([" xx", "x x", "xx "], 2),

  ([ " o", "o ", " x", "x "], 0)
  ]

_samples :: [Sample]
_samples = concatMap fst _pricedSamples

getSamplePrice :: Sample -> Price
getSamplePrice s = head [p | (ss, p) <- _pricedSamples, s `elem` ss]
--------------------------------------

-- getPricedSteps [((1, 5), (4, 2), "x x ")]  ->  [((2,4), 100),((4,2), 100)]
getPricedSteps :: [(Pos, Pos, Sample)] -> [(Price, Pos)]
getPricedSteps trios = (reverse . sort) (concatMap getPricedStep trios)
 where
  getPricedStep :: (Pos, Pos, Sample) -> [(Price, Pos)]
  getPricedStep (p1, p2, sample) =
    [(getSamplePrice sample, p) | (p, s) <- zip (interpolation (p1, p2)) sample, s == ' ']



---------------------------------------------

findSamples :: Matr -> [(Pos, Pos, Sample)]
findSamples matr = concatMap f _samples
 where
   f sample  = (\(p1, p2) -> (p1, p2, sample)) <$> findSample matr sample

findSample :: Matr -> Sample -> [(Pos, Pos)]
findSample matr xs = let
  n = length xs
  val r c = matr !! (r * _size + c)

  hor = [horEq r c | r <- [0.._size-1], c <- [0.._size-n] ]
  ver = [verEq r c | r <- [0.._size-n], c <- [0.._size-1] ]
  dia = [diaEq r c | r <- [0.._size-n], c <- [0.._size-n] ]
  aid = [aidEq r c | r <- [0.._size-n], c <- [n-1.._size-1] ]

  horEq r c = if [val r (c+i)     | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r, c+n-1))
  verEq r c = if [val (r+i) c     | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r+n-1, c))
  diaEq r c = if [val (r+i) (c+i) | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r+n-1, c+n-1))
  aidEq r c = if [val (r+i) (c-i) | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r+n-1, c-n+1))
 in
  [ fromJust x |  x <- hor ++ ver ++ dia ++ aid, isJust x]



isCellEmpty :: Table -> Pos -> IO Bool
isCellEmpty table (r, c) = do
  if r >= 0 && r < _size && c >= 0 && c < _size
    then do
      v <- get table (r, c)
      return $ v == ' '
    else
      return False

whoWon :: Table -> IO (Val, (Pos, Pos))  
whoWon t = do
   matr <- getElems t
   let xs = findSample matr "xxxxx"
   let os = findSample matr "ooooo"
   if xs /= []
   then return ('x', head xs)
   else if os /= []
   then return ('o', head os)
   else return (' ', ((0,0), (0,0)))
-------------------------------------------------------------
{-- 
  знаходимо всі зразки, які є в даному стані гри  
  
  знаходимо всі доцільні ходи "о"
  для кожного ходу 
      знаходимо все доцільні ходи "х"
      оцінюємо стан після кожного ходу "х"
  
  

--}

nextStep :: Val -> Table -> Int -> IO (Price, Pos)
nextStep v t 0 = do
  steps <- getBestSteps t v
  if null steps
    then rndStepO t
    else return $ head steps

nextStep v t level = do
  steps <- getBestSteps t v
  if null steps
    then rndStepO t
    else do
      -- ts :: IO [Table] - список таблиц со сделанными ходами 
      ts <- mapM (\(_, pos) -> do {t' <- mapArray id t; put t' pos 'о'; return t'} ) steps 
      let v' = if v == 'x' then 'o' else 'x'
      let steps' = mapM (\tbl -> nextStep v' tbl (level-1)) ts

      return $ head steps



getBestSteps :: Table -> Val -> IO [(Price, Pos)]
getBestSteps t 'o' = getBestStepsO t
getBestSteps t 'x' = do
    steps <- getBestStepsO t 
    let steps' = map (\(price, pos) -> (f price, pos)) steps
    return steps'
 where 
    f 1000 = 999
    f 999 = 1000
    f 100 = 99
    f 99 = 100
    f 50 = 49
    f 49 = 50
    f x = x
tgetBestSteps t _ = error ""

getBestStepsO :: Table -> IO [(Price, Pos)]
getBestStepsO t = do
  matr <- getElems t
  let steps = getPricedSteps (findSamples matr)
  return $ take 5 steps




rndStepO :: Table -> IO (Price, Pos)
rndStepO t = do
   g0 <- initStdGen
   let (a, g1) = uniformR (2, 7) g0 :: (Int, StdGen)
   let (b, g) = uniformR (2, 7) g1 :: (Int, StdGen)
   empty <- isCellEmpty t (a, b)
   if empty
     then return (0, (a, b))
     else rndStepO t







src/Draw.hs
module Draw (drawTable, hilightWin, hilightPos) where
import Data.Array.IO ( getElems, readArray, writeArray, MArray(newArray), IOArray )
import Consul
import System.IO ( hFlush, stdout )
import Control.Concurrent ( threadDelay )
import Lib2
_cross = "><"
_zero = "<>"

drawTable :: Table -> IO ()
drawTable table = do
  putStr $ hideCur ++ rc 0 0
  cs <-  getElems table
  sequence_ [drawCell cs r c | r <- [0.._size - 1], c <- [0.._size - 1]]
  putStrLn ""

drawCell :: [Char] -> Int -> Int -> IO ()
drawCell cs row col
  | c == '\n' = putStrLn (rc row' col')
  | c == 'x'  = putStr $ rc row' (col'+1) ++ green  ++ _cross
  | c == 'o'  = putStr $ rc row' (col'+1) ++ red ++ _zero
  | otherwise = putStr $ rc row' (col'+1) ++ gray ++ show row ++ show col
  where
    i = _size * row + col
    c = cs !! i
    col' = (col + 1) * 3
    row' = row + 2

hilightWin :: Table -> (Char, (Pos, Pos)) -> IO ()
hilightWin table (who, segment) = do
   mapM_ f (interpolation segment)
 where
   simbol = if who == 'x' then _cross else _zero
   f (r, c) = putStr $ rc (r + 2) (c*3 + 4) ++ yellow  ++ simbol

hilightPos :: Table -> Char -> Pos -> IO ()
hilightPos t who (r, c) = do
  putStr $ rc r' c' ++ hideCur ++ colorB ++ color ++ sym
  threadDelay 200000
  putStr $ rc r' c' ++ blackB ++ color ++ sym 
 where 
    r' = r + 2
    c' = c * 3 + 4
    (color, colorB, sym) = if who == 'x' 
       then (red, redB, _cross) 
       else (green, greenB, _zero)

     




src/Lib2.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Use putStrLn" #-}
{-# HLINT ignore "Eta reduce" #-}
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
{-# HLINT ignore "Use mapM_" #-}
{-# HLINT ignore "Use first" #-}
{-# HLINT ignore "Use tuple-section" #-}

module Lib2 (
  _size, interpolation, put, puts,  newTable,
  priceTableAfterStep, selStepsOnTable, findSamplesS, findSamplesT, 
  randomStep, nextSteps, isCellEmpty, whoWon,
  Table, Pos,  Matr, Sample, Price, Val, Level
 ) where

import Control.Monad (sequence_, mapM_)
import Data.Array.IO --( getElems, readArray, writeArray, MArray(newArray), IOArray )
import Data.Foldable (Foldable(foldl'))
import Data.Time.Clock ( getCurrentTime, UTCTime(utctDayTime) )
import Data.Maybe ( fromJust, isJust )
import System.Random ( initStdGen, uniformR, StdGen )
import Consul
import System.IO ( hFlush, stdout )
import Control.Concurrent ( threadDelay )
import Data.List (sort, nub)


type Table = IOArray Int Val
type Pos = (Int, Int)
type Matr = [Val]
type Sample = [Val]
type Price = Int
type Val = Char    -- 'x', 'o', ' '
type Level = Int   -- 0, 1, 2, 3
_size = 10

_WIDTH = 5

get :: Table -> Pos -> IO Val
get table (r, c) = readArray table (r * _size + c)

put :: Table -> Pos -> Val -> IO ()
put table (r, c) = writeArray table (r * _size + c)

puts :: Table -> [Int] -> Val -> IO ()
puts table ns v = mapM_ (\n -> put table (divMod n 10) v) ns

newTable :: IO Table
newTable = newArray (0, _size^2 - 1) ' '

isCellEmpty :: Table -> Pos -> IO Bool
isCellEmpty table (r, c) = do
  if r >= 0 && r < _size && c >= 0 && c < _size
    then do
      v <- get table (r, c)
      return $ v == ' '
    else
      return False

whoWon :: Table -> IO (Val, (Pos, Pos))
whoWon t = do
   matr <- getElems t
   let xs = findSample matr "xxxxx"
   let os = findSample matr "ooooo"
   if xs /= []
    then return ('x', head xs)
   else if os /= []
    then return ('o', head os)
    else return (' ', ((0,0), (0,0)))


-- UTILS ------------------------------------------------

interpolation :: (Pos, Pos) -> [Pos]
interpolation ((r1, c1), (r2, c2))
  | r1 == r2  = [(r1, c) | c <- [c1..c2] ]
  | c1 == c2  = [(r, c1) | r <- [r1..r2] ]
  | otherwise = [(r, c)  | (r, c) <- zip [r1..r2] [c1, c1 + dCol ..c2]]
  where
    dCol = signum (c2 - c1)

insertSepByN n [] sep = []
insertSepByN n xs sep = let (zs, rs) = splitAt n xs
               in zs ++ [sep] ++ insertSepByN n rs sep

swapXO cs = map f cs where
  f 'x' = 'o'
  f 'o' = 'x'
  f c = c

------------ образцы -------------------------------------------

-- samplesOs -- для оценки перспективности хода 'o'
_samplesOs :: [([Sample], Price)]
_samplesOs = [ 
  ([" oooo", "o ooo", "oo oo", "ooo o", "oooo "], 10),
  ([" xxxx", "x xxx", "xx xx", "xxx x", "xxxx "], 9),

  ([" ooo", "o oo", "oo o", "ooo "], 8),
  ([" xxx", "x xx", "xx x", "xxx "], 7),
  ([" oo", "o o", "oo "], 6),
  ([" xx", "x x", "xx "], 5),
  ([" o", "o "], 4),
  ([" x", "x "], 3)
  ]
-- samplesOs -- для оценки таблицы после хода 'o'
_samplesOt :: [([Sample], Price)]
_samplesOt = [
  (["ooooo"], 10^6),   -- 'o' win in 0 step
  ([" xxxx", "x xxx", "xx xx", "xxx x", "xxxx "], -10^5), -- 'x' win in 1 step 
  ([" oooo "], 10^4),   -- 'o' win in 0 step
  
  ([" oooo", "o ooo", "oo oo", "ooo o", "oooo "], 1000),   -- 'o' win in 1 step

  (["  xxx", " x xx", " xx x", " xxx ", "x  xx", "x x x", "x xx ", "xx x ", "xxx  " ], -100), -- 'x' win in 2 steps  
  (["  ooo", " o oo", " oo o", " ooo ", "o  oo", "o o o", "o oo ", "oo o ", "ooo  " ], 100),   -- 'o' win in 2 steps

  (["xx   ", "x x  ", "x  x ", "x   x", " xx  ", " x x ", " x  x", "  x x", "   xx" ], -10),  -- 'x' win in 3 steps  
  (["oo   ", "o o  ", "o  o ", "o   o", " oo  ", " o o ", " o  o", "  o o", "   oo" ], 10)     -- 'o' win in 3 steps  
  ]

_samplesS :: Val -> [(Sample, Price)]
_samplesS val = let
  f (ss, p) = [(s, p) | s <- ss] :: [(Sample, Price)]
  pairs = concatMap f _samplesOs
 in
  if val == 'o'
   then pairs
   else map (\(s, p) -> (swapXO s, p)) pairs

_samplesT :: Val -> [(Sample, Price)]
_samplesT val = let
  f (ss, p) = [(s, p) | s <- ss] :: [(Sample, Price)]
  pairs = concatMap f _samplesOt
 in
  if val == 'o'
   then pairs
   else map (\(s, p) -> (swapXO s, p)) pairs

----------- поиск образцов -------------------------

findSamplesS v matr = findSamples v (_samplesS v) matr
findSamplesT v matr = findSamples v (_samplesT v) matr

findSamples :: Val -> [(Sample, Price)] -> Matr -> [((Pos, Pos), Sample, Price)]
findSamples v samples matr = concatMap f samples
 where
   f :: (Sample, Price) -> [((Pos, Pos), Sample, Price)]
   f (sample, price)  = fmap (\pp -> (pp, sample, price)) (findSample matr sample)

findSample :: Matr -> Sample -> [(Pos, Pos)]
findSample matr xs = let
  n = length xs
  val r c = matr !! (r * _size + c)

  hor = [horEq r c | r <- [0.._size-1], c <- [0.._size-n] ]
  ver = [verEq r c | r <- [0.._size-n], c <- [0.._size-1] ]
  dia = [diaEq r c | r <- [0.._size-n], c <- [0.._size-n] ]
  aid = [aidEq r c | r <- [0.._size-n], c <- [n-1.._size-1] ]

  horEq r c = if [val r (c+i)     | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r, c+n-1))
  verEq r c = if [val (r+i) c     | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r+n-1, c))
  diaEq r c = if [val (r+i) (c+i) | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r+n-1, c+n-1))
  aidEq r c = if [val (r+i) (c-i) | i <- [0..n-1]] /= xs then Nothing
    else Just ((r, c), (r+n-1, c-n+1))
 in
  [ fromJust x |  x <- hor ++ ver ++ dia ++ aid, isJust x]

--------- цена таблицы после сделанного хода -------------------

priceTableAfterStep :: Val -> Table -> IO Price
priceTableAfterStep v t = do
  matr <- getElems t
  let ps = findSamples v (_samplesT v) matr
  let price = sum [p | (pp, s, p) <- ps]
  return price

--------- выбор походящих ходов из заданной таблицы ------------------

selStepsOnTable :: Val -> Table -> IO [(Price, Pos)]
selStepsOnTable v t = do
  matr <- getElems t
  let samples = findSamples v (_samplesS v) matr  :: [((Pos, Pos), Sample, Price)]
  if null samples
    then randomStep v t
    else return $ getPricedSteps samples

getPricedSteps :: [((Pos, Pos), Sample, Price)] -> [(Price, Pos)]
getPricedSteps ps = (take _WIDTH . reverse . sort . nub) (concatMap pricedStep ps)
  where
    pricedStep :: ((Pos, Pos), Sample, Price) -> [(Price, Pos)]
    pricedStep ((pos1, pos2), sample, price) =
      [(abs price, pos) | (pos, s) <- zip (interpolation (pos1, pos2)) sample, s == ' ']

--------------------------------------------------------------
randomStep :: Val -> Table -> IO [(Price, Pos)]
randomStep v t = do
   g0 <- initStdGen
   let (a, g1) = uniformR (2, 7) g0 :: (Int, StdGen)
   let (b, g) = uniformR (2, 7) g1 :: (Int, StdGen)
   empty <- isCellEmpty t (a, b)
   if empty
     then return [(0, (a, b))]
     else randomStep v t

-- несколько вариантов лучшего хода игрока
nextSteps :: Val -> Table -> Level -> IO [(Price, Pos)]
nextSteps v table 0 = selStepsOnTable v table


-- несколько вариантов лучшего хода игрока
nextSteps v table level = do
  steps <- selStepsOnTable v table
  if fst (head steps) == 10000 
   then return steps            -- есть выигрышный ход
   else do 
    let v' = if v == 'o' then 'x' else 'o'
    --__pwp ("the best "++ [v] ++" steps") steps 
    -- tables - список таблиц со сделанными ходами 
    tables <- mapM (getNextTables v table) steps  :: IO [Table]
    -- самый лучший ход противника из таблицы tbl
    let contrSteps tbl  = head <$> nextSteps v' tbl (level-1) ::  IO (Price, Pos)
    -- лучшие ходы противника из всех таблиц tables
    steps' <- mapM contrSteps tables :: IO [(Price, Pos)]
    --__pwp "the best contra steps" steps' 
    -- соединяем цены ходов противника с ходами игрока
    let steps'' = zipWith (\(p, pos) (p', pos') -> (p', pos)) steps steps' ::  [(Price, Pos)]
    --__pwp "new prices" steps'' 
    -- выбираем 5 худших  
    let steps''' = sort steps''
    return steps'''

__pwp prompt x =  print (prompt ++ "  " ++ show x ++ "    >") >> getLine

getNextTables :: Val -> Table -> (Price, Pos) -> IO Table
getNextTables v table (_, pos) = do 
  t <- mapArray id table 
  put t pos v; 
  return t

src/Consul.hs
-- https://ru.wikipedia.org/wiki/%D0%A3%D0%BF%D1%80%D0%B0%D0%B2%D0%BB%D1%8F%D1%8E%D1%89%D0%B8%D0%B5_%D0%BF%D0%BE%D1%81%D0%BB%D0%B5%D0%B4%D0%BE%D0%B2%D0%B0%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D0%BE%D1%81%D1%82%D0%B8_ANSI#Windows_%D0%B8_DOS
module Consul where

clrscr = "\27[0;0H\27[J"
origin =  "\27[0;0H"
hideCur = "\27[?25l" -- CSI ?25l
showCur = "\27[?25h" -- CSI ?25h

-- text colors
black = "\27[30m"
red = "\27[31m"
green = "\27[32m"
yellow = "\27[33m"
blue = "\27[34m"
magenta = "\27[35m"
cian = "\27[36m"
white = "\27[37m"
norm = "\27[m"
gray = "\27[1m\27[30m"
-- back color
blackB = "\27[40m"
redB = "\27[41m"
greenB = "\27[42m"
yellowB = "\27[43m"
blueB = "\27[44m"
magentaB = "\27[45m"
cianB = "\27[46m"
whiteB = "\27[47m"


-- cur movement
rc row col = "\27["++ show row ++";"++ show col ++"f"  -- row,col = 1,2,3...
up n = "\27["++ show n ++"A"
down n = "\27["++ show n ++"B"
forward n = "\27["++ show n ++"C"
back n = "\27["++ show n ++"D"





app/Main.hs
module Main where


import Data.Array.IO ()
import Control.Monad ( when )
import Text.Read ( readMaybe )
import Lib2
import Consul
import Draw
import System.IO ( hFlush, stdout )

main :: IO ()
main = do
  t <- newTable
  -- puts t [01,03,05,07] 'x'
  -- puts t [45,55, 65,75] 'o'
  putStr clrscr
  drawTable t 
  run t

-------------------------------------------------------
run :: Table -> IO ()
run t = do
  putStr $ norm ++ rc 12 0 ++ showCur ++ "q-quit >"
  hFlush stdout

  s <- getLine
  when ('q' `notElem` s)  (do
      let nn = readMaybe s :: Maybe Int
      case nn of
        Nothing -> run t
        Just x -> twoMoves t x
       )

twoMoves t nn = do
  let posX = divMod nn 10
  isEmpty <- isCellEmpty t posX
  if not isEmpty
    then run t
    else do
      put t posX 'x'
      hilightPos t 'x' posX   
      who <- whoWon t
      if fst who == 'x'
        then epilog t who
        else do
          poses <- nextSteps 'o' t 2  -- <<<<<<<<<<<<<<<<<
          let posO = snd $ head poses
          put t posO 'o'
          hilightPos t 'o' posO  
          who <- whoWon t
          if fst who == 'o'
            then epilog t who
            else run t

epilog t who = do
  hilightWin t who
  putStr $ norm ++ rc 12 0 ++ showCur ++  "Continue ? [y], n >"
  hFlush stdout
  ans <- getLine
  when (ans /= "n") main



